pkg/readiness/pruner/pruner_test.go

```diff
@@ -0,0 +1,270 @@
+
+	for _, tt := range tts {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx, cancelFunc := context.WithCancel(context.Background())
+
+			require.NoError(t, testutils.ApplyFixtures(tt.fixturesPath, cfg), "applying base fixtures")
+
+			testRes := setupTest(ctx, t, testOptions{addConstrollers: tt.startControllers})
+
+			testutils.StartManager(ctx, t, testRes.manager)
+
+			require.Eventually(t, func() bool {
+				return testRes.expectationsPruner.tracker.Populated()
+			}, timeout, tick, "waiting on tracker to populate")
+
+			for _, name := range tt.syncsetsToDelete {
+				u := &unstructured.Unstructured{}
+				u.SetGroupVersionKind(syncsetGVK)
+				u.SetName(name)
+
+				require.NoError(t, testRes.k8sClient.Delete(ctx, u), fmt.Sprintf("deleting syncset %s", name))
+			}
+			if tt.deleteConfig != "" {
+				u := &unstructured.Unstructured{}
+				u.SetGroupVersionKind(configGVK)
+				u.SetNamespace("gatekeeper-system")
+				u.SetName(tt.deleteConfig)
+
+				require.NoError(t, testRes.k8sClient.Delete(ctx, u), fmt.Sprintf("deleting config %s", tt.deleteConfig))
+			}
+
+			testRes.expectationsPruner.pruneNotWatchedGVKs()
+
+			require.Eventually(t, func() bool {
+				return testRes.expectationsPruner.tracker.Satisfied()
+			}, timeout, tick, "waiting on tracker to get satisfied")
+
+			cancelFunc()
+		})
+	}
+}
+
+// Test_ExpectationsMgr_missedInformers verifies that the pruner can handle a scenario
+// where the readiness tracker's state will never match the informer cache events.
+func Test_ExpectationsMgr_missedInformers(t *testing.T) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	// Set up one data store for readyTracker:
+	// we will use a separate lister for the tracker from the mgr client and make
+	// the contents of the readiness tracker be a superset of the contents of the mgr's client
+	// the syncset will look like:
+	// *fakes.SyncSetFor("syncset-a", []schema.GroupVersionKind{podGVK, configMapGVK}),
+	testRes := setupTest(ctx, t, testOptions{testLister: makeTestLister(t), addExpectationsPruner: true, addConstrollers: true})
+
+	// Set up another store for the controllers and watchManager
+	syncsetA := fakes.SyncSetFor("syncset-a", []schema.GroupVersionKind{podGVK})
+	require.NoError(t, testRes.k8sClient.Create(ctx, syncsetA))
+
+	testutils.StartManager(ctx, t, testRes.manager)
+
+	require.Eventually(t, func() bool {
+		return testRes.expectationsPruner.tracker.SyncSourcesSatisfied()
+	}, timeout, tick, "waiting on sync sources to get satisfied")
+
+	// As configMapGVK is absent from this syncset-a, the CacheManager will never observe configMapGVK
+	// being deleted and will never cancel the data expectation for configMapGVK.
+	require.NoError(t, testRes.k8sClient.Delete(ctx, syncsetA))
+
+	require.Eventually(t, func() bool {
+		return testRes.expectationsPruner.tracker.Satisfied()
+	}, timeout, tick, "waiting on tracker to get satisfied")
+
+	cancelFunc()
+}
+
+func makeTestLister(t *testing.T) readiness.Lister {
```

Comment:
`makeTestLister` should not exist.

By moving the set of items to-be-created outside of the test, readers of the test lose context as to what the state of the system is expected to be.

Because the only thing this function does is pre-bake expectations, we should remove it.